components/derive/src/db_traits.rs

```diff
@@ -116,3 +120,310 @@ pub fn impl_object_hash(input: TokenStream) -> TokenStream {
+    }
+
+    fn root_constructor(&self, field_index: usize) -> impl ToTokens {
+        let from_access = quote!(exonum_merkledb::access::FromAccess);
+        let ident = self.ident(field_index);
+        let name = &self.name_suffix;
+        quote!(#ident: #from_access::from_access(access.clone(), #name.into())?)
+    }
+}
+
+impl FromAccess {
+    fn access_fn(&self) -> impl ToTokens {
+        let fn_impl = if self.attrs.transparent {
+            let from_access = quote!(exonum_merkledb::access::FromAccess);
+            let ident = self.fields[0].ident(0);
+            quote!(Ok(Self { #ident: #from_access::from_access(access, addr)? }))
+        } else {
+            let field_constructors = self
+                .fields
+                .iter()
+                .enumerate()
+                .map(|(i, field)| field.constructor(i));
+            quote!(Ok(Self { #(#field_constructors,)* }))
+        };
+
+        let access_ident = &self.access_ident;
+        quote! {
+            fn from_access(
+                access: #access_ident,
+                addr: exonum_merkledb::IndexAddress,
+            ) -> Result<Self, exonum_merkledb::access::AccessError> {
+                #fn_impl
+            }
+        }
+    }
+
+    fn root_fn(&self) -> impl ToTokens {
+        let fn_impl = if self.attrs.transparent {
+            let from_access = quote!(exonum_merkledb::access::FromAccess);
+            let ident = self.fields[0].ident(0);
+            quote!(Ok(Self { #ident: #from_access::from_root(access)? }))
+        } else {
+            let field_constructors = self
+                .fields
+                .iter()
+                .enumerate()
+                .map(|(i, field)| field.root_constructor(i));
+            quote!(Ok(Self { #(#field_constructors,)* }))
+        };
+
+        let access_ident = &self.access_ident;
+        quote! {
+            fn from_root(
+                access: #access_ident,
+            ) -> Result<Self, exonum_merkledb::access::AccessError> {
+                #fn_impl
+            }
+        }
+    }
+}
+
+impl ToTokens for FromAccess {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let name = &self.ident;
+        let tr = quote!(exonum_merkledb::access::FromAccess);
+        let access_ident = &self.access_ident;
+        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
+
+        let from_access_fn = self.access_fn();
+        let from_root_fn = self.root_fn();
+
+        let expanded = quote! {
+            impl #impl_generics #tr<#access_ident> for #name #ty_generics #where_clause {
+                #from_access_fn
+                #from_root_fn
+            }
+        };
+        tokens.extend(expanded);
+
+        if self.attrs.is_schema() {
+            let expanded = quote! {
+                impl #impl_generics #name #ty_generics #where_clause {
```

Comment:
Perhaps we might to introduce a new trait like `FromAccess` but for constructing schemas like `IntoSchema` to avoid of ad-hoc `new` method implementation. In my opinion this approach is more generic and a little bit closer to `derive` philoshophy.

Reply:
Or even more: we can use `FromAccess` for both cases.

```rust

impl<T> FromAccess for Schema<T> { ... }

// and

impl<T, A> FromAccess for SomeIndex<T> for (T, A) where A: Into<IndexAddress> { .. }
```

Reply:
Then this trait would need to be imported into the scopes that want to use `Schema::new()`. It's possible to introduce `new()` in `FromAccess` or in an extension trait to `FromAccess`. Would any of this make sense to you? 🙃

Also, regarding `(T, A) where A: Into<IndexAddress>`: does this assume removing `from_access` method from `FromAccess`? I'm not sure this is going to work with the transparent representation, or would be composable in general.